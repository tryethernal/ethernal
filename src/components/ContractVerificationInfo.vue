<template>
    <div>
        <v-card outlined class="mb-4">
            <v-card-text>
                <v-row>
                    <v-col cols="6"><b>Contract Name:</b> {{ contract.verification.contractName }}</v-col>
                    <v-col cols="6">
                        <b>Optimizer:</b> <span v-if="contract.verification.runs"><b>Yes</b> with <b>{{ contract.verification.runs }}</b> runs</span>
                        <span v-else>No</span>
                    </v-col>
                </v-row>
                <v-row>
                    <v-col cols="6"><b>Compiler Version:</b> {{ contract.verification.compilerVersion }}</v-col>
                    <v-col cols="6"><b>EVM Version:</b> {{ contract.verification.evmVersion }}</v-col>
                </v-row>
                <v-divider class="my-6"></v-divider>
                <v-row v-for="(source, idx) in contract.verification.sources" :key="idx">
                    <v-col>
                        <h5>File {{ idx + 1}} of {{ contract.verification.sources.length }}: {{ source.fileName }}</h5>
                        <editor class="editor" :ref="`editor-${idx}`" v-model="source.content" @init="editorInit" lang="solidity" theme="chrome" height="500"></editor>
                    </v-col>
                </v-row>
            </v-card-text>
        </v-card>

        <v-card v-if="displayConstructorArguments" outlined class="mb-6">
            <v-card-title>
                Constructor Arguments
                <v-spacer></v-spacer>
                <span class="text-caption">
                    <a :class="{ underlined: formattedConstructorArguments }" @click="formattedConstructorArguments = true">Formatted</a> | <a :class="{ underlined: !formattedConstructorArguments }" @click="formattedConstructorArguments = false">Raw</a>
                </span>
            </v-card-title>
            <v-card-text>
                <template v-if="formattedConstructorArguments">
                    <div v-for="(arg, idx) in decodedConstructorArguments" :key="idx" style="white-space: pre;">
                        <Formatted-Sol-Var :input="arg" :notInteractive="true" :value="arg.value" />
                    </div>
                </template>
                <span v-else>{{ zeroXifiedConstructorArguments }}</span>
            </v-card-text>
        </v-card>

        <v-card v-if="displayLibraries" outlined class="mb-6">
            <v-card-title>Libraries</v-card-title>
            <v-card-text>
                <div v-for="(libraryName, idx) in Object.keys(contract.verification.libraries)" :key="idx">
                    {{ libraryName }} => {{ contract.verification.libraries[libraryName] }}
                </div>
            </v-card-text>
        </v-card>
    </div>
</template>

<script>
const ethers = require('ethers');
const editor = require('vue2-ace-editor');
import FormattedSolVar from './FormattedSolVar';

export default {
    name: 'ContractVerificationInfo',
    props: ['contract'],
    components: {
        editor,
        FormattedSolVar
    },
    data: () => ({
        formattedConstructorArguments: true
    }),
    mounted() {
        for (let i = 0; i < this.contract.verification.sources.length; i++) {
            const editor = this.$refs[`editor-${i}`][0].editor;
            editor.setOptions({
                readOnly: true,
                showPrintMargin: false,
                useWorker: false,
                maxLines: 25
            });
        }
    },
    methods: {
        editorInit: function () {
            require('brace/ext/language_tools')
            require('brace/theme/chrome')
        }
    },
    computed: {
        zeroXifiedConstructorArguments() {
            return this.contract.verification.constructorArguments.startsWith('0x') ?
                this.contract.verification.constructorArguments :
                `0x${this.contract.verification.constructorArguments}`;
        },
        decodedConstructorArguments() {
            const iface = new ethers.utils.Interface(this.contract.abi);
            const constructorInputs = JSON.parse(iface.deploy.format(ethers.utils.FormatTypes.json)).inputs;

            /*
                This won't handle well tuples in tuples, but hopefully it'll be good enough for now.
                I'd say that I'll improve later, but we all know it's probably never going to happen
            */
            const decodedInputs = ethers.utils.defaultAbiCoder.decode(iface.deploy.inputs.map(i => {
                if (i.type == 'tuple')
                    return `tuple(${i.components.map(c => c.type).join(',')})`;
                else
                    return i.type;
            }), this.zeroXifiedConstructorArguments);
            const decoded = [];
            for (let i = 0; i < decodedInputs.length; i++) {
                decoded.push({
                    ...constructorInputs[i],
                    value: decodedInputs[i]
                });
            }
            return decoded;
        },
        displayConstructorArguments() {
            return this.contract.verification && this.contract.verification.constructorArguments;
        },
        displayLibraries() {
            return this.contract.verification && Object.keys(this.contract.verification.libraries).length > 0;
        }
    }
}
</script>
<style scoped style="scss">
.editor {
    border: 1px solid var(--v-background-base);
    border-radius: 5px;
}
.underlined {
    text-decoration: underline;
}
</style>
