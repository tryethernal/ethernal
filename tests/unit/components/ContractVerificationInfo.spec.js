import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

jest.mock('@metamask/detect-provider');
import ContractVerificationInfo from '@/components/ContractVerificationInfo.vue';

let helper;
const stubs = [
    'Formatted-Sol-Var'
];

describe('ContractVerificationInfo.vue', () => {

    beforeEach(() => {
        jest.clearAllMocks();
        helper = new MockHelper();
    });

    it('Should display code / constructor arguments / libraries', async () => {
        const wrapper = helper.mountFn(ContractVerificationInfo, {
            propsData: {
                contract: {
                    bytecode: '0x60',
                    asm: 'abc',
                    abi: require('../fixtures/NFTCollectibleABI.json'),
                    verificationStatus: 'success',
                    verification: {
                        contractName: 'My Contract',
                        evmVersion: 'london',
                        compilerVersion: 'compiler',
                        runs: 200,
                        constructorArguments: '0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000046162636400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000005000000000000000000000000000000000000000000000000000000000000000461626364000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000001610000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016200000000000000000000000000000000000000000000000000000000000000',
                        libraries: {
                            Library: '0x12345'
                        },
                        sources: [{ fileName: 'File.sol', content: 'i am solidity code' }]
                    }
                }
            },
            stubs: stubs,
            getters: {
                isPublicExplorer: jest.fn(() => true)
            }
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
