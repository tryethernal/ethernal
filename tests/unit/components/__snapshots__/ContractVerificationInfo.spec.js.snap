// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractVerificationInfo.vue Should display code / constructor arguments / libraries 1`] = `
<div>
  <div class="mb-4 v-card v-sheet v-sheet--outlined theme--light">
    <div class="v-card__text">
      <div class="row">
        <div class="col col-6"><b>Contract Name:</b> My Contract</div>
        <div class="col col-6"><b>Optimizer:</b> <span><b>Yes</b> with <b>200</b> runs</span></div>
      </div>
      <div class="row">
        <div class="col col-6"><b>Compiler Version:</b> compiler</div>
        <div class="col col-6"><b>EVM Version:</b> london</div>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="my-6 v-divider theme--light">
      <div class="row">
        <div class="col">
          <h5>File 1 of 1: File.sol</h5>
          <div style="height: 500px; width: 100%" class="editor ace_editor ace-chrome"><textarea class="ace_text-input" wrap="off" autocorrect="off" autocapitalize="off" spellcheck="false" style="opacity: 0;"></textarea>
            <div class="ace_gutter" aria-hidden="true">
              <div class="ace_layer ace_gutter-layer ace_folding-enabled"></div>
              <div class="ace_gutter-active-line"></div>
            </div>
            <div class="ace_scroller">
              <div class="ace_content">
                <div class="ace_layer ace_print-margin-layer">
                  <div class="ace_print-margin" style="left: 4px; visibility: hidden;"></div>
                </div>
                <div class="ace_layer ace_marker-layer"></div>
                <div class="ace_layer ace_text-layer" style="padding: 0px 4px;"></div>
                <div class="ace_layer ace_marker-layer"></div>
                <div class="ace_layer ace_cursor-layer ace_hidden-cursors">
                  <div class="ace_cursor"></div>
                </div>
              </div>
            </div>
            <div class="ace_scrollbar ace_scrollbar-v" style="display: none; width: 20px;">
              <div class="ace_scrollbar-inner" style="width: 20px;"></div>
            </div>
            <div class="ace_scrollbar ace_scrollbar-h" style="display: none; height: 20px;">
              <div class="ace_scrollbar-inner" style="height: 20px;"></div>
            </div>
            <div style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;">
              <div style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"></div>
              <div style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;">XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="mb-6 v-card v-sheet v-sheet--outlined theme--light">
    <div class="v-card__title">
      Constructor Arguments
      <div class="spacer"></div> <span class="text-caption"><a class="underlined">Formatted</a> | <a class="">Raw</a></span>
    </div>
    <div class="v-card__text">
      <div style="white-space: pre;">
        <formatted-sol-var-stub notinteractive="true" value="abcd" input="[object Object]"></formatted-sol-var-stub>
      </div>
      <div style="white-space: pre;">
        <formatted-sol-var-stub notinteractive="true" value="abcd,a,b,true,5" input="[object Object]"></formatted-sol-var-stub>
      </div>
    </div>
  </div>
  <div class="mb-6 v-card v-sheet v-sheet--outlined theme--light">
    <div class="v-card__title">Libraries</div>
    <div class="v-card__text">
      <div>
        Library =&gt; 0x12345
      </div>
    </div>
  </div>
</div>
`;
